File: packages/twerk-slots/src/twerk_slots/cli/slot/checkout.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Annotated, Any

import click

from twerk_core import get_console
from twerk_core.clinkr.exit import ClinkrExit
from twerk_core.clinkr.operation import clinkr_operation
from twerk_slots.allocation import (
    DetachedHeadError,
    DirtyCurrentWorktreeError,
    PoolFullError,
    SlotAllocationError,
    SlotAllocationResult,
    allocate_slot_for_branch,
    allocate_slot_for_current_branch,
)
from twerk_slots.cli.slot.context import load_slots_context
from twerk_slots.context import SlotsCliContext
from twerk_slots.gateway.clipboard import ClipboardCopySuccess
from twerk_slots.naming import extract_slot_number
from twerk_slots.repo_context import NoRepoSentinel, ensure_slots_metadata_dir


@dataclass(frozen=True)
class SlotCheckoutRequest:
    branch_name: Annotated[
        str | None,
        click.Argument(["branch_name"], required=False, default=None),
    ] = None
    new_branch: Annotated[
        bool,
        click.Option(["-b", "--new"], "new_branch", is_flag=True, default=False),
    ] = False
    current: Annotated[bool, click.Option(["--current"], is_flag=True, default=False)] = False
    no_clipboard: Annotated[
        bool,
        click.Option(["--no-clipboard"], is_flag=True, default=False),
    ] = False


@dataclass(frozen=True)
class SlotCheckoutResult:
    slot_name: str
    branch_name: str
    worktree_path: str
    cd_command: str
    already_assigned: bool
    created_branch: bool
    evicted_slot: str | None
    current_wt_note: str | None
    clipboard_copied: bool
    clipboard_skipped: bool
    clipboard_failure_reason: str | None
    clipboard_failure_detail: str | None

    def to_json_dict(self) -> dict[str, Any]:
        return {
            "slot_name": self.slot_name,
            "branch_name": self.branch_name,
            "worktree_path": self.worktree_path,
            "cd_command": self.cd_command,
            "already_assigned": self.already_assigned,
            "created_branch": self.created_branch,
            "evicted_slot": self.evicted_slot,
            "current_wt_note": self.current_wt_note,
            "clipboard_copied": self.clipboard_copied,
            "clipboard_skipped": self.clipboard_skipped,
            "clipboard_failure_reason": self.clipboard_failure_reason,
            "clipboard_failure_detail": self.clipboard_failure_detail,
        }


def render_slot_checkout(result: SlotCheckoutResult) -> None:
    console = get_console()
    if result.evicted_slot is not None:
        console.print(
            f"[yellow]Evicted {result.evicted_slot} to make room for {result.branch_name}[/yellow]"
        )
    if result.current_wt_note is not None:
        console.print(f"[yellow]{result.current_wt_note}[/yellow]")
    if result.already_assigned:
        if extract_slot_number(result.slot_name) is None:
            console.print(
                f"[dim]{result.branch_name}[/dim] is already checked out in the "
                f"main worktree at [bold cyan]{result.worktree_path}[/bold cyan]"
            )
        else:
            console.print(
                f"[dim]{result.branch_name}[/dim] is already assigned to "
                f"[bold cyan]{result.slot_name}[/bold cyan]"
            )
    else:
        console.print(
            f"Checked out [bold cyan]{result.slot_name}[/bold cyan] -> "
            f"[green]{result.branch_name}[/green]"
        )
    click.echo(result.cd_command)
    if result.clipboard_skipped:
        return
    if result.clipboard_copied:
        console.print("[dim]Copied cd command to clipboard.[/dim]")
    else:
        detail = result.clipboard_failure_detail or "pbcopy failed"
        console.print(f"[dim]Clipboard unavailable ({detail})[/dim]")


def _pool_full_failure(outcome: PoolFullError) -> ClinkrExit[SlotCheckoutResult]:
    return ClinkrExit[SlotCheckoutResult].failure(
        error_type="pool_full",
        message=(
            f"Pool is full. Oldest slot {outcome.oldest_slot} holds "
            f"'{outcome.oldest_branch}'. Free a slot before checking out a new branch."
        ),
    )


def _build_result(
    ctx: SlotsCliContext,
    allocation: SlotAllocationResult,
    *,
    created_branch: bool,
    current_wt_note: str | None,
    no_clipboard: bool,
) -> SlotCheckoutResult:
    worktree_path = str(allocation.worktree_path)
    cd_command = f"cd {worktree_path}"
    clipboard_copied = False
    clipboard_failure_reason: str | None = None
    clipboard_failure_detail: str | None = None
    if not no_clipboard:
        outcome = ctx.clipboard.copy(cd_command)
        if isinstance(outcome, ClipboardCopySuccess):
            clipboard_copied = True
        else:
            clipboard_failure_reason = outcome.reason
            clipboard_failure_detail = outcome.detail
    return SlotCheckoutResult(
        slot_name=allocation.slot_name,
        branch_name=allocation.branch_name,
        worktree_path=worktree_path,
        cd_command=cd_command,
        already_assigned=allocation.already_assigned,
        created_branch=created_branch,
        evicted_slot=allocation.evicted_slot,
        current_wt_note=current_wt_note,
        clipboard_copied=clipboard_copied,
        clipboard_skipped=no_clipboard,
        clipboard_failure_reason=clipboard_failure_reason,
        clipboard_failure_detail=clipboard_failure_detail,
    )


@clinkr_operation(
    name="checkout",
    help="Check out a branch into a pool slot worktree (like `git checkout [-b]`).",
    aliases=("co",),
    human_renderer=render_slot_checkout,
)
def run_checkout_slot(
    ctx: click.Context, request: SlotCheckoutRequest
) -> ClinkrExit[SlotCheckoutResult]:
    inputs_provided = sum((request.branch_name is not None, request.current))
    if inputs_provided > 1:
        return ClinkrExit[SlotCheckoutResult].failure(
            error_type="mutually_exclusive_args",
            message="Pass exactly one of BRANCH_NAME or --current.",
        )
    if inputs_provided == 0:
        return ClinkrExit[SlotCheckoutResult].failure(
            error_type="missing_arg",
            message="Pass BRANCH_NAME or --current to identify the branch.",
        )
    if request.current and request.new_branch:
        return ClinkrExit[SlotCheckoutResult].failure(
            error_type="mutually_exclusive_args",
            message="-b/--new cannot be combined with --current.",
        )

    slots_ctx = load_slots_context(ctx)
    if isinstance(slots_ctx, NoRepoSentinel):
        return ClinkrExit[SlotCheckoutResult].failure(
            error_type="not_in_repo", message=slots_ctx.message
        )

    ensure_slots_metadata_dir(slots_ctx.repo, slots_ctx.storage)
    now = datetime.now(UTC).isoformat()

    if request.current:
        try:
            current_outcome = allocate_slot_for_current_branch(
                slots_ctx, cwd=slots_ctx.repo.root, now=now, force=False
            )
        except SlotAllocationError as exc:
            return ClinkrExit[SlotCheckoutResult].failure(
                error_type="slot_allocation_error", message=str(exc)
            )
        if isinstance(current_outcome, DetachedHeadError):
            return ClinkrExit[SlotCheckoutResult].failure(
                error_type="detached_head",
                message=(
                    f"HEAD at {current_outcome.cwd} is detached. Check out a branch "
                    f"before running `slot checkout --current`."
                ),
            )
        if isinstance(current_outcome, DirtyCurrentWorktreeError):
            return ClinkrExit[SlotCheckoutResult].failure(
                error_type="dirty_worktree",
                message=(
                    f"Current worktree at {current_outcome.cwd} has uncommitted changes. "
                    f"Commit or stash before running `slot checkout --current`."
                ),
            )
        if isinstance(current_outcome, PoolFullError):
            return _pool_full_failure(current_outcome)
        return ClinkrExit.ok(
            _build_result(
                slots_ctx,
                current_outcome.allocation,
                created_branch=False,
                current_wt_note=current_outcome.current_wt_note,
                no_clipboard=request.no_clipboard,
            )
        )

    assert request.branch_name is not None  # validated above
    branch_name = request.branch_name
    branch_exists = slots_ctx.git.branch_exists(branch_name)
    created_branch = False

    if request.new_branch:
        if branch_exists:
            return ClinkrExit[SlotCheckoutResult].failure(
                error_type="branch_exists",
                message=(
                    f"Branch '{branch_name}' already exists. "
                    f"Drop -b to check out the existing branch."
                ),
            )
        slots_ctx.git.create_branch(branch_name, "HEAD", force=False)
        created_branch = True
    elif not branch_exists:
        return ClinkrExit[SlotCheckoutResult].failure(
            error_type="branch_missing",
            message=(
                f"Branch '{branch_name}' does not exist. Pass -b/--new to create it from HEAD."
            ),
        )

    outcome = allocate_slot_for_branch(
        slots_ctx,
        branch_name=branch_name,
        now=now,
        force=False,
    )
    if isinstance(outcome, PoolFullError):
        return _pool_full_failure(outcome)
    return ClinkrExit.ok(
        _build_result(
            slots_ctx,
            outcome,
            created_branch=created_branch,
            current_wt_note=None,
            no_clipboard=request.no_clipboard,
        )
    )

File: ts/packages/capabilities/slot/src/lifecycle/checkout.ts
import type { RepoSlotContext, SlotCliContext } from "../context.ts";
import { buildSlotInventory } from "../inventory.ts";
import { planCheckout, planCurrentCheckout, type CheckoutPlan } from "../planning.ts";
import { ensureSlotsMetadataDir } from "../repo-context.ts";
import {
	branchInUseFailure,
	poolFullFailure,
	type LifecycleResult,
	type SlotLifecycleFailure,
} from "./common.ts";
import { executeCurrentWorktreeRedirect } from "./current-worktree-redirect.ts";

export interface SlotCheckoutOutcome {
	slotName: string;
	branchName: string;
	worktreePath: string;
	alreadyAssigned: boolean;
	createdBranch: boolean;
	currentWtNote: string | null;
}

export type SlotCheckoutResult = LifecycleResult<SlotCheckoutOutcome>;

type ExecutableCheckoutPlan = Extract<
	CheckoutPlan,
	{ type: "reuse_assignment" | "branch_in_main_worktree" | "assign_to_slot" }
>;

export async function checkoutBranch(
	ctx: SlotCliContext,
	branchName: string,
	options: { shouldCreateBranch: boolean; base: string | null },
): Promise<SlotCheckoutResult> {
	if (ctx.repo.type !== "repo")
		return {
			type: "failure",
			failure: { errorType: ctx.repo.errorType, message: ctx.repo.message },
		};
	const repoCtx: RepoSlotContext = { ...ctx, repo: ctx.repo };
	await ensureSlotsMetadataDir(repoCtx.repo, repoCtx.storage);

	const branchExists = await repoCtx.git.branchExists(branchName);
	let hasCreatedBranch = false;
	if (options.shouldCreateBranch) {
		if (branchExists)
			return failure(
				"branch-exists",
				`Branch '${branchName}' already exists. Drop -b to check out the existing branch.`,
			);
		if (options.base !== null && !(await repoCtx.git.branchExists(options.base)))
			return failure("base-missing", `Base branch '${options.base}' does not exist.`);
		const createFailure = await repoCtx.git.createBranch(branchName, options.base ?? "HEAD", {
			shouldForce: false,
		});
		if (createFailure !== null)
			return failure(
				"branch-create-failed",
				`Failed to create branch '${branchName}': ${createFailure.message}`,
			);
		hasCreatedBranch = true;
	} else if (!branchExists) {
		return failure(
			"branch-missing",
			`Branch '${branchName}' does not exist. Pass -b/--new to create it from HEAD.`,
		);
	}

	const inventory = await buildSlotInventory(repoCtx.git, {
		mainRepoRoot: repoCtx.repo.mainRepoRoot,
	});
	const plan = await planCheckout(inventory, repoCtx.git, branchName);
	if (plan.type === "pool_full")
		return {
			type: "failure",
			failure: poolFullFailure(plan.assigned, { action: "checking out a new branch" }),
		};
	if (plan.type === "branch_in_use")
		return { type: "failure", failure: branchInUseFailure(plan.occupancy) };
	return await executeCheckoutPlan(plan, repoCtx, {
		branchName,
		hasCreatedBranch,
		currentWtNote: null,
	});
}

export async function checkoutCurrent(ctx: SlotCliContext): Promise<SlotCheckoutResult> {
	if (ctx.repo.type !== "repo")
		return {
			type: "failure",
			failure: { errorType: ctx.repo.errorType, message: ctx.repo.message },
		};
	const repoCtx: RepoSlotContext = { ...ctx, repo: ctx.repo };
	await ensureSlotsMetadataDir(repoCtx.repo, repoCtx.storage);

	const currentPlan = await planCurrentCheckout(repoCtx.git, {
		cwd: repoCtx.repo.root,
		mainRepoRoot: repoCtx.repo.mainRepoRoot,
	});
	if (currentPlan.type === "allocation_failure")
		return failure("slot-allocation-error", currentPlan.message);
	if (currentPlan.type === "detached_head")
		return failure(
			"detached-head",
			`HEAD at ${currentPlan.cwd} is detached. Check out a branch before running \`sdl slot checkout --current\`.`,
		);
	if (currentPlan.type === "dirty_worktree")
		return failure(
			"dirty-worktree",
			`Current worktree at ${currentPlan.cwd} has uncommitted changes. Commit or stash before running \`sdl slot checkout --current\`.`,
		);
	if (currentPlan.plan.type === "pool_full")
		return {
			type: "failure",
			failure: poolFullFailure(currentPlan.plan.assigned, { action: "checking out a new branch" }),
		};
	if (currentPlan.plan.type === "branch_in_use")
		return { type: "failure", failure: branchInUseFailure(currentPlan.plan.occupancy) };
	if (currentPlan.redirect !== null) {
		const redirectFailure = await executeCurrentWorktreeRedirect(currentPlan.redirect, repoCtx);
		if (redirectFailure !== null) return { type: "failure", failure: redirectFailure };
	}
	return await executeCheckoutPlan(currentPlan.plan, repoCtx, {
		branchName: currentPlan.branchName,
		hasCreatedBranch: false,
		currentWtNote: currentPlan.currentWtNote,
	});
}

async function executeCheckoutPlan(
	plan: ExecutableCheckoutPlan,
	ctx: RepoSlotContext,
	options: { branchName: string; hasCreatedBranch: boolean; currentWtNote: string | null },
): Promise<SlotCheckoutResult> {
	if (plan.type === "reuse_assignment") {
		return ok({
			slotName: plan.record.slotName,
			branchName: options.branchName,
			worktreePath: plan.record.path,
			alreadyAssigned: true,
			createdBranch: options.hasCreatedBranch,
			currentWtNote: options.currentWtNote,
		});
	}
	if (plan.type === "branch_in_main_worktree") {
		return ok({
			slotName: "",
			branchName: options.branchName,
			worktreePath: plan.mainPath,
			alreadyAssigned: true,
			createdBranch: options.hasCreatedBranch,
			currentWtNote: options.currentWtNote,
		});
	}
	const checkoutFailure = await ctx.git.checkoutBranch(plan.record.path, options.branchName);
	if (checkoutFailure !== null)
		return failure(
			"checkout-failed",
			`Failed to check out '${options.branchName}' into ${plan.record.slotName}: ${checkoutFailure.message}`,
		);
	return ok({
		slotName: plan.record.slotName,
		branchName: options.branchName,
		worktreePath: plan.record.path,
		alreadyAssigned: false,
		createdBranch: options.hasCreatedBranch,
		currentWtNote: options.currentWtNote,
	});
}

function ok(outcome: SlotCheckoutOutcome): SlotCheckoutResult {
	return { type: "ok", outcome };
}

function failure(errorType: string, message: string): SlotCheckoutResult {
	const lifecycleFailure: SlotLifecycleFailure = { errorType: errorType, message };
	return { type: "failure", failure: lifecycleFailure };
}
